contrib/platform/src/com/sun/jna/platform/win32/Winspool.java

```diff
@@ -508,6 +512,20 @@ public boolean hasAttribute(int value) {
         }
     }
 
+    /**
+     * The PRINTER_INFO_3 structure specifies printer security information.
+     */
+    @FieldOrder({"pSecurityDescriptor"})
+    public static class PRINTER_INFO_3 extends Structure {
+        public WinNT.SECURITY_DESCRIPTOR_RELATIVE pSecurityDescriptor;
```

Comment:
Although `WinNT` provided a plain `SECURITY_DESCRIPTOR`, the struct is malformed for use with `PRINTER_INFO_3` and will throw a runtime error [`java.lang.IllegalStateException: Array fields must be initialized`](https://stackoverflow.com/questions/43018371/exception-in-thread-main-java-lang-illegalstateexception-array-fields-must-be)`.

Fortunately `SECURITY_DESCRIPTOR_RELATIVE` matches the struct design exactly that's returned by `PRINTER_INFO_3`, fixing this runtime exception.

Reply:
This is a pointer to `SECURITY_DESCRIPTOR` and I believe the `_RELATIVE` version is the correct "by reference" mapping.

Reply:
Reopening... so going back on the `DWORD` vs `int` conversation, I'm looking for examples and found that `PRINTER_INFO_2` uses `public INT_PTR pSecurityDescriptor;`.  Should I switch `PRINTER_INFO_3` to match?  If so, how would I go about getting the data, a manual cast?

Reply:
Same issue occurs for `PRINTER_INFO_2`, which returns a `INT_PTR` in place of the `pDevMode`, whereas the newly written `PRINTER_INFO_8`, `PRINTER_INFO_9` both use this PR's struct.   `pDevMode` appears again in `L_PRINTER_DEFAULTS`, but this time as a `Pointer` object.

I can fix these, but those using the AP in its current form  will break.

I tried to convert the `INT_PTR` to a struct, but I got an error `java.lang.UnsupportedOperationException: This pointer is opaque: const@0x1a31bba904c`.


```java
// Doesn't work :(
public DEVMODE(Pointer p) {
    super(p);
    ensureAllocated();
    read();
}
```

Reply:
> Reopening... so going back on the `DWORD` vs `int` conversation, I'm looking for examples and found that `PRINTER_INFO_2` uses `public INT_PTR pSecurityDescriptor;` however this PR uses the `WinNT.SECURITY_DESCRIPTOR_RELATIVE`. Should I switch `PRINTER_INFO_3` to match? 

My apologies, I answered the first one thinking about a completely different recent change.   Ultimately you need to map a pointer of some sort, and then use that pointer to map to the structure it points to.   Using the `ByReference` tag on the structure accomplishes this transparently and should probably be your first choice.  So I'd suggest `WinNT.SECURITY_DESCRIPTOR_RELATIVE.ByReference`.   Check the commit dates on the other structure, it's entirely possible it was committed first.

Don't change old ones to break compatibility.  If you want consistency, feel free to match them, but `INT_PTR` is a pointer-sized integer so you'd need to use `toPointer()` to convert it and then use _that_ pointer to pass to a constructor.

> It seems less useful, but I'd like to be consistent. If so, how would I go about getting the data, construct a `Pointer` from the int and manually cast it? If that's the case, is this better than just using the correct struct to begin with?

I think the structure (by reference) is the easiest.    So you get `INT_PTR ip`, you do `Pointer p = ip.toPointer()` and then pass `p` to the pointer constructor of the structure, essentially this (`FooStructure extends Structure`):
```java
FooStructure(Pointer p) {
  super(p);
  // anything else you need to do
}
```